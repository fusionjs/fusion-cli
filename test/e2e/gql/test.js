// @flow
/* eslint-env node */

const t = require('assert');
const path = require('path');
const request = require('request-promise');

const puppeteer = require('puppeteer');

const {cmd, start} = require('../utils.js');

const runnerPath = require.resolve('../../../bin/cli-runner');
const countTests = require('../test-jest-app/fixture/src/count-tests');

const {promisify} = require('util');
const exec = promisify(require('child_process').exec);

const dev = require('../setup.js');

const dir = path.resolve(__dirname, './fixture');

test('`fusion dev` works with gql', async () => {
  const app = dev(dir);
  await app.setup();
  const url = app.url();
  try {
    expect(await request(`${url}/schema`)).toMatchSnapshot();
    const page = await app.browser().newPage();
    await page.goto(`${url}/`, {waitUntil: 'load'});
    const browserSchema = await page.evaluate(() => {
      return typeof window !== undefined && window.schema; //eslint-disable-line
    });
    expect(browserSchema).toMatchSnapshot();
  } catch (e) {
    t.ifError(e);
  }
  app.teardown();
}, 100000);

test('`fusion build --production` works with gql', async () => {
  let browser;
  await await cmd(`build --dir=${dir} --production`);
  const {proc, port} = await start(`--dir=${dir}`, {
    env: Object.assign({}, process.env, {NODE_ENV: 'production'}),
  });
  try {
    expect(await request(`http://localhost:${port}/schema`)).toMatchSnapshot();
    browser = await puppeteer.launch({
      args: ['--no-sandbox', '--disable-setuid-sandbox'],
    });
    const page = await browser.newPage();
    await page.goto(`http://localhost:${port}/`, {waitUntil: 'load'});
    const browserSchema = await page.evaluate(() => {
      return typeof window !== undefined && window.schema; //eslint-disable-line
    });
    expect(browserSchema).toMatchSnapshot();
  } catch (e) {
    t.ifError(e);
  }
  // $FlowFixMe
  browser.close();
  proc.kill();
}, 100000);

test('`fusion test` with gql macro', async () => {
  const args = `test --dir=${dir} --configPath=../../../../build/jest/jest-config.js`;

  const cmd = `require('${runnerPath}').run('node ${runnerPath} ${args}')`;
  const response = await exec(`node -e "${cmd}"`);
  t.equal(countTests(response.stderr), 2, 'ran 2 tests');
});

test('`fusion test` coverage with gql', async () => {
  const args = `test --dir=${dir} --configPath=../../../../build/jest/jest-config.js --coverage`;

  const cmd = `require('${runnerPath}').run('node ${runnerPath} ${args}')`;
  const response = await exec(`node -e "${cmd}"`);
  t.equal(countTests(response.stderr), 2, 'ran 2 tests');
});
