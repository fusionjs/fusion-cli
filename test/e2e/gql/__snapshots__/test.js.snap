// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`fusion build --production\` works with gql 1`] = `"{\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"ObjectTypeDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Query\\"},\\"interfaces\\":[],\\"directives\\":[],\\"fields\\":[{\\"kind\\":\\"FieldDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"user\\"},\\"arguments\\":[],\\"type\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"User\\"}},\\"directives\\":[]}]},{\\"kind\\":\\"ObjectTypeDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"User\\"},\\"interfaces\\":[],\\"directives\\":[],\\"fields\\":[{\\"kind\\":\\"FieldDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"firstName\\"},\\"arguments\\":[],\\"type\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"String\\"}},\\"directives\\":[]}]}],\\"loc\\":{\\"start\\":0,\\"end\\":62,\\"source\\":{\\"body\\":\\"type Query {\\\\n  user: User\\\\n}\\\\n\\\\ntype User {\\\\n  firstName: String\\\\n}\\",\\"name\\":\\"GraphQL request\\",\\"locationOffset\\":{\\"line\\":1,\\"column\\":1}}}}"`;

exports[`\`fusion build --production\` works with gql 2`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "user",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "User",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Query",
      },
    },
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "firstName",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "User",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 62,
    "source": Object {
      "body": "type Query {
  user: User
}

type User {
  firstName: String
}",
      "locationOffset": Object {
        "column": 1,
        "line": 1,
      },
      "name": "GraphQL request",
    },
    "start": 0,
  },
}
`;

exports[`\`fusion dev\` works with gql 1`] = `"{\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"ObjectTypeDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Query\\"},\\"interfaces\\":[],\\"directives\\":[],\\"fields\\":[{\\"kind\\":\\"FieldDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"user\\"},\\"arguments\\":[],\\"type\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"User\\"}},\\"directives\\":[]}]},{\\"kind\\":\\"ObjectTypeDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"User\\"},\\"interfaces\\":[],\\"directives\\":[],\\"fields\\":[{\\"kind\\":\\"FieldDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"firstName\\"},\\"arguments\\":[],\\"type\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"String\\"}},\\"directives\\":[]}]}],\\"loc\\":{\\"start\\":0,\\"end\\":62,\\"source\\":{\\"body\\":\\"type Query {\\\\n  user: User\\\\n}\\\\n\\\\ntype User {\\\\n  firstName: String\\\\n}\\",\\"name\\":\\"GraphQL request\\",\\"locationOffset\\":{\\"line\\":1,\\"column\\":1}}}}"`;

exports[`\`fusion dev\` works with gql 2`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "user",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "User",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Query",
      },
    },
    Object {
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "firstName",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "User",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 62,
    "source": Object {
      "body": "type Query {
  user: User
}

type User {
  firstName: String
}",
      "locationOffset": Object {
        "column": 1,
        "line": 1,
      },
      "name": "GraphQL request",
    },
    "start": 0,
  },
}
`;
